### ISSUE TITLE
[Console] Using an OutputFormatter leads to distortedly rendered Tables

### ISSUE THREAD

**Issue by k0pernikus** [symfony/symfony#15517]
This issue stems from my [stackoverflow question about how to indent write operations on a ConsoleInterface](http://stackoverflow.com/questions/31946977). I am wondering if this may be an issue (or possible) feature request with the Console Component concerning how the OutputInterface and the `Table` helper work together when an `OutputFormatter` is applied.

My use case is: Indenting the write operations to an `OutputInterface` as the rendered information has nested context.

I also write a bunch of tables on the `OutputInterface` with the `Table` helper, so I was expecting to do something like this:

``` php
new Table($output); // Output can be BufferedOutput or ConsoleOutput, any OutputInterface
$output->writeln('0. run');
$someTable->render();
$output->increaseIndentLevel(); // pseudocode
$output->writeln('1. run');
$someTable->render();
```

which should render like so:

```
0. run
+---------------+-----------------------+------------------+
| ISBN          | Title                 | Author           |
+---------------+-----------------------+------------------+
| 99921-58-10-7 | Divine Comedy         | Dante Alighieri  |
| 9971-5-0210-0 | A Tale of Two Cities  | Charles Dickens  |
| 960-425-059-0 | The Lord of the Rings | J. R. R. Tolkien |
+---------------+-----------------------+------------------+
     1. run
     +---------------+-----------------------+------------------+
     | ISBN          | Title                 | Author           |
     +---------------+-----------------------+------------------+
     | 99921-58-10-7 | Divine Comedy         | Dante Alighieri  |
     | 9971-5-0210-0 | A Tale of Two Cities  | Charles Dickens  |
     | 960-425-059-0 | The Lord of the Rings | J. R. R. Tolkien |
     +---------------+-----------------------+------------------+
```

I tried solving it with an `OutputFormatter`:

``` php
<?php
namespace Hive\App;

use Symfony\Component\Console\Formatter\OutputFormatter;
use Symfony\Component\Console\Formatter\OutputFormatterStyleInterface;
use Symfony\Component\Console\Output\OutputInterface;

/**
 * IndentedOutputFormatter
 **/
class IndentedOutputFormatter extends OutputFormatter
{
    const INDENT_AMOUNT = 4;

    private $indentLevel = 0;

    /**
     * Formats a message according to the given styles.
     * @param string $message The message to style
     * @return string The styled message
     * @api
     */
    public function format($message)
    {
        $message = parent::format($message);
        if ($this->indentLevel === 0) {
            return $message;
        }

        $amount = self::INDENT_AMOUNT * $this->indentLevel;
        $prependBy = str_repeat(' ', $amount);
        $message = $prependBy . $message;

        return $message;
    }

    /**
     *
     */
    public function increaseLevel()
    {
        $this->indentLevel = $this->indentLevel + 1;
    }

    /**
     *
     */
    public function decreaseLevel()
    {
        $this->indentLevel = $this->indentLevel - 1;
    }
}
```

And use it with:

``` php
/**
 * @param InputInterface  $input
 * @param OutputInterface $output
 * @return int
 */
protected function execute(InputInterface $input, OutputInterface $output)
{
    $headers = [
        'ISBN',
        'Title',
        'Author',
    ];

    $rows = [
        [
            '99921-58-10-7',
            'Divine Comedy',
            'Dante Alighieri',

        ],
        [
            '9971-5-0210-0',
            'A Tale of Two Cities',
            'Charles Dickens',

        ],
        [
            '960-425-059-0',
            'The Lord of the Rings',
            'J. R. R. Tolkien',
        ],
    ];


    $formatter = new IndentedOutputFormatter();
    $output->setFormatter($formatter);
    $table = new Table($output);
    $table->setHeaders($headers);
    $table->setRows($rows);

    foreach (range(0, 3) as $currentRun) {
        $output->writeln("$currentRun. run");
        $formatter->increaseLevel();
        $table->render();
    }

    return 0;
}
```

Yet the formatter also applies on the table cells, distorting the Output.

```
0. run
    +-------------------+---------------------------+----------------------+
    |     ISBN              |     Title                     |     Author               |    
    +-------------------+---------------------------+----------------------+
    |     99921-58-10-7     |     Divine Comedy             |     Dante Alighieri      |    
    |     9971-5-0210-0     |     A Tale of Two Cities      |     Charles Dickens      |    
    |     960-425-059-0     |     The Lord of the Rings     |     J. R. R. Tolkien     |    
    +-------------------+---------------------------+----------------------+
    1. run
        +-----------------------+-------------------------------+--------------------------+
        |         ISBN                  |         Title                         |         Author                   |        
        +-----------------------+-------------------------------+--------------------------+
        |         99921-58-10-7         |         Divine Comedy                 |         Dante Alighieri          |        
        |         9971-5-0210-0         |         A Tale of Two Cities          |         Charles Dickens          |        
        |         960-425-059-0         |         The Lord of the Rings         |         J. R. R. Tolkien         |        
        +-----------------------+-------------------------------+--------------------------+
        2. run
            +---------------------------+-----------------------------------+------------------------------+
            |             ISBN                      |             Title                             |             Author                       |            
            +---------------------------+-----------------------------------+------------------------------+
            |             99921-58-10-7             |             Divine Comedy                     |             Dante Alighieri              |            
            |             9971-5-0210-0             |             A Tale of Two Cities              |             Charles Dickens              |            
            |             960-425-059-0             |             The Lord of the Rings             |             J. R. R. Tolkien             |            
            +---------------------------+-----------------------------------+------------------------------+
            3. run
                +-------------------------------+---------------------------------------+----------------------------------+
                |                 ISBN                          |                 Title                                 |                 Author                           |                
                +-------------------------------+---------------------------------------+----------------------------------+
                |                 99921-58-10-7                 |                 Divine Comedy                         |                 Dante Alighieri                  |                
                |                 9971-5-0210-0                 |                 A Tale of Two Cities                  |                 Charles Dickens                  |                
                |                 960-425-059-0                 |                 The Lord of the Rings                 |                 J. R. R. Tolkien                 |                
                +-------------------------------+---------------------------------------+----------------------------------+
```

I am wondering if this is an issue in how the Table helper renders the table or in how I try to implement the indent functionality.

Basically: Is this an issue with the Console Component or is it me using it wrong? If it is expected behavior could we turn it into some kind of feature request? As I am sure that indenting is an reappearing use case.

I am open to provide a pull request, I just need some feedback first :)

**Comment by apfelbox** [comment]
@k0pernikus you never call `decreaseLevel()`, you only ever `increase`?


```php
// ...
foreach (range(0, 3) as $currentRun) {
    $output->writeln("$currentRun. run");
    $formatter->increaseLevel();
    $table->render();
    $formatter-> decreaseLevel(); // <-- ?
}
```

**Comment by k0pernikus** [comment]
@apfelbox Yes, by choice for this example to showcase the odd behavior, namely that the indentation is applied for the content of the table cells as well.

It makes sense though, as the Output gets injected in the Table helper, yet there, the OutputFormatter should not be applied. It should only applied at the line level, not on the rendering at a table cell.

Basically, this is about how to control or set OutputFormatter from different contexts. This OutputFormatter should not apply when called from inside a Table helper for instance.

**Comment by stof** [comment]
@k0pernikus your issue is that you assume that OutputFormatter deals with full lines, and so you can add indentation during formatting. This assumption is totally wrong. The OutputFormatter deals with any string we want to format, not with lines (and it could even have to deal with multiline messages btw).

**Comment by k0pernikus** [comment]
@stof Where would you see the responsibility on formattting the (multi)line level?

**Comment by stof** [comment]
@k0pernikus I think you could achieve it with a custom OutputInterface implementation, which would decorate the real one and indent lines before calling the real one. you would then pass this OutputInterface to the Table

**Comment by javiereguiluz** [comment]
I'm closing this as "won't fix" because this doesn't look like a common enough feature to include it in Symfony core. In addition to the proposal made by @stof, you can solve this using a BufferedOutput to get the entire table rendering and then easily apply an indentation before actually displaying it (here https://github.com/symfony/symfony-demo/blob/master/src/AppBundle/Command/ListUsersCommand.php you can see an example of BufferedOutput + tables in commands). Thanks!
